joomla/joomla-cms
# [4.1] Making Google happy about breadcrumbs again.

This PR fixes structured data errors that made Google flag breadcrumbs as invalid.

When a breadcrumb item was a menu heading, instead of a link, Google would flag it as an invalid breadcrumb because it did not have a link. This is fixed by removing the JSON markup for items without a link, with the exception of the last item/current page.

This pull request fixes issues #37508 and #37510.

Basically reproducing my changes from my prior 4.0 pull request, but updated for the new JSON format.

### Summary of Changes

1. Added a check if a link exists in a breadcrumb item. If not, skip it and don't add it to the JSON.
2. Added the last item (aka current page) to the JSON without a link. This is valid per [Google SEO guidelines](https://developers.google.com/search/docs/advanced/structured-data/breadcrumb#list-item); they will pull the current page URL for the last item.

### Testing Instructions

Create a site menu with a menu heading and subpages, like so:

- Home
- Menu Heading
  - Subpage 1

Create a breadcrumbs module. Ensure the module is published and visible on Subpage 1.

Navigate to subpage 1.

If this site is accessible over the internet:

Copy the URL and paste it into the field here: https://search.google.com/test/rich-results

If this site is hosted locally:

View source of Subpage 1 and copy it into the code box on https://search.google.com/test/rich-results
(Alternatively, you can find the generated JSON for the module and just copy that into the code box.)

### Actual result BEFORE applying this Pull Request

Google detects the breadcrumbs but shows an error complaining about the missing URL in an item.

### Expected result AFTER applying this Pull Request

Google correctly detects the breadcrumbs and marks them as valid.

### Documentation Changes Required

None as far as I am aware.

### Additional Notes

I intentionally am not checking if the module is set to display the last item or not because for the purposes of the JSON, it is better to include it and does not affect the front end display of the module.
